ably/rest_client.go

```diff
@@ -500,155 +497,49 @@ func (c *REST) do(ctx context.Context, r *request) (*http.Response, error) {
 	return c.doWithHandle(ctx, r, c.handleResponse)
 }
 
-// fallbackCache this caches a successful fallback host for 10 minutes.
-type fallbackCache struct {
-	running  bool
-	host     string
-	duration time.Duration
-	cancel   func()
-	mu       sync.RWMutex
-}
+func (c *REST) doWithHandle(ctx context.Context, r *request, handle func(*http.Response, interface{}) (*http.Response, error)) (*http.Response, error) {
+	defer c.hosts.resetVisitedFallbackHosts()
 
-func (f *fallbackCache) get() string {
-	if f.isRunning() {
-		f.mu.RLock()
-		h := f.host
-		f.mu.RUnlock()
-		return h
+	maxHTTPRequestLimit := c.opts.HTTPMaxRetryCount
+	if maxHTTPRequestLimit == 0 {
+		maxHTTPRequestLimit = defaultOptions.HTTPMaxRetryCount
 	}
-	return ""
-}
-
-func (f *fallbackCache) isRunning() bool {
-	f.mu.RLock()
-	v := f.running
-	f.mu.RUnlock()
-	return v
-}
 
-func (f *fallbackCache) run(host string) {
-	f.mu.Lock()
-	now := time.Now()
-	duration := defaultOptions.FallbackRetryTimeout // spec RSC15f
-	if f.duration != 0 {
-		duration = f.duration
-	}
-	ctx, cancel := context.WithDeadline(context.Background(), now.Add(duration))
-	f.running = true
-	f.host = host
-	f.cancel = cancel
-	f.mu.Unlock()
-	<-ctx.Done()
-	f.mu.Lock()
-	f.running = false
-	f.mu.Unlock()
-}
+	host := c.hosts.getPreferredHost()
+	iteration := 0
 
-func (f *fallbackCache) stop() {
-	f.cancel()
-	// we make sure we have stopped
 	for {
-		if !f.isRunning() {
-			return
+		req, err := c.newHTTPRequest(ctx, r, host)
+		if err != nil {
+			return nil, err
 		}
-	}
-}
-
-func (f *fallbackCache) put(host string) {
-	if f.get() != host {
-		if f.isRunning() {
-			f.stop()
+		if host != c.hosts.getPrimaryHost() { // set hostheader for fallback host
+			req.Host = req.URL.Host // RSC15j set host header, https://github.com/golang/go/issues/7682, since req.Host overrides req.URL.Host, use the same value
+			req.Header.Set(hostHeader, host)
```

Comment:
Won't `newHTTPRequest` do this anyway?

Reply:
No it will only set `req.URL.Host` and not `req.Host`

Reply:
From the docs on [`http.Request.Host`](https://pkg.go.dev/net/http#Request):

```
// For client requests, Host optionally overrides the Host
// header to send. If empty, the Request.Write method uses
// the value of URL.Host. Host may contain an international
// domain name.
Host string
```

So it claims to do exactly what we're doing here (i.e. set `Host` to the value of `URL.Host`), are you sure this is required?

Reply:
It's required to set host header explicitly, I think we discussed this before. Although, as per documentation things are not clear about it.

Reply:
I have updated it, but I strongly remember not able to set header when Host was not set : (